src/formats/jpeg.zig

```diff
@@ -0,0 +1,1084 @@
+
+        if (start_of_spectral_selection > 63) {
+            return error.InvalidSpectralSelectionValue;
+        }
+
+        if (end_of_spectral_selection < start_of_spectral_selection or end_of_spectral_selection > 63) {
+            return error.InvalidSpectralSelectionValue;
+        }
+
+        // If Ss = 0, then Se = 63.
+        if (start_of_spectral_selection == 0 and end_of_spectral_selection != 63) {
+            return error.InvalidSpectralSelectionValue;
+        }
+
+        segment_size -= 2;
+        if (JPEG_VERY_DEBUG) std.debug.print("  Spectral selection: {}-{}\n", .{ start_of_spectral_selection, end_of_spectral_selection });
+
+        const approximation_bits = try reader.readByte();
+        const approximation_high = @intCast(u4, approximation_bits >> 4);
+        const approximation_low = @intCast(u4, approximation_bits & 0b1111);
+
+        segment_size -= 1;
+        if (JPEG_VERY_DEBUG) std.debug.print("  Approximation bit position: high={} low={}\n", .{ approximation_high, approximation_low });
+
+        std.debug.assert(segment_size == 0);
+
+        return ScanHeader{
+            .components = components,
+            .start_of_spectral_selection = start_of_spectral_selection,
+            .end_of_spectral_selection = end_of_spectral_selection,
+            .approximation_high = approximation_high,
+            .approximation_low = approximation_low,
+        };
+    }
+};
+
+// See figure A.6 in T.81.
+const ZigzagOffsets = blk: {
+    var offsets: [64]usize = undefined;
+    offsets[0] = 0;
+
+    var current_offset: usize = 0;
+    var direction: enum { NorthEast, SouthWest } = .NorthEast;
+    var i: usize = 1;
+    while (i < 64) : (i += 1) {
+        switch (direction) {
+            .NorthEast => {
+                if (current_offset < 8) {
+                    // Hit top edge
+                    current_offset += 1;
+                    direction = .SouthWest;
+                } else if (current_offset % 8 == 7) {
+                    // Hit right edge
+                    current_offset += 8;
+                    direction = .SouthWest;
+                } else {
+                    current_offset -= 7;
+                }
+            },
+            .SouthWest => {
+                if (current_offset >= 56) {
+                    // Hit bottom edge
+                    current_offset += 1;
+                    direction = .NorthEast;
+                } else if (current_offset % 8 == 0) {
+                    // Hit left edge
+                    current_offset += 8;
+                    direction = .NorthEast;
+                } else {
+                    current_offset += 7;
+                }
+            },
+        }
+
+        if (current_offset >= 64) {
+            @compileError(std.fmt.comptimePrint("ZigzagOffsets: Hit offset {} (>= 64) at index {}!\n", .{ current_offset, i }));
+        }
+
+        offsets[i] = current_offset;
+    }
+
+    break :blk offsets;
+};
```

Comment:
This is one of the killer features of Zig :)

Reply:
Yeah, I love it, same with the DCT tables below. It's so much nicer to actually embed the algorithm instead of obfuscated lookup tables :^)